networkit/cpp/community/LouvainMapEquation.cpp

```diff
@@ -0,0 +1,583 @@
+
+        // correct the cut for potentially moved neighbors
+        // the already calculated cut updates assumed none of them were moved
+        G->forEdgesOf(u, [&](node, node v, edgeweight weight) {
+            if (u >= v)
+                return;
+
+            const index originClusterOfNeighbor = result[v],
+                        targetClusterOfNeighbor = nextPartition[v];
+            if (originClusterOfNeighbor != targetClusterOfNeighbor) {
+                const double w2 = 2 * weight;
+
+                if (originCluster == originClusterOfNeighbor) {
+                    originClusterCutUpdate -= w2;
+                } else if (targetClusterOfNeighbor == originCluster) {
+                    originClusterCutUpdate += w2;
+                }
+
+                if (targetClusterOfNeighbor == targetCluster) {
+                    targetClusterCutUpdate -= w2;
+                } else if (originClusterOfNeighbor == targetCluster) {
+                    targetClusterCutUpdate += w2;
+                }
+            }
+        });
+
+// apply cut updates
+#pragma omp atomic
+        clusterCut[originCluster] += originClusterCutUpdate;
+#pragma omp atomic
+        clusterCut[targetCluster] += targetClusterCutUpdate;
+
+        totalCutUpdate += originClusterCutUpdate + targetClusterCutUpdate;
+    }
+
+#pragma omp atomic
+    totalCut += totalCutUpdate;
+}
+
+// for every node. store its neighbors that are in the current chunk, and their old cluster IDs, and
+// edge weights
+bool LouvainMapEquation::tryLocalMove(node u, SparseVector<double> &neighborClusterWeights,
+                                      /* SLM specifics */
+                                      std::vector<Move> &moves, bool synchronous) {
+    // find neighbor clusters
+    double vol = 0;
+    double loop = 0;
+    double weightToCurrent = 0;
+    const index currentCluster = result[u];
+
+    G->forEdgesOf(u, [&](node, node v, edgeweight weight) {
+        vol += weight;
+        if (u != v) {
+            const index neighborCluster = result[v];
+            if (neighborCluster == currentCluster) {
+                weightToCurrent += weight;
+            } else {
+                if (!neighborClusterWeights.indexIsUsed(neighborCluster))
+                    neighborClusterWeights.insert(neighborCluster, 0);
+                neighborClusterWeights[neighborCluster] += weight;
+            }
+        } else {
+            loop += weight;
+            vol += weight;
+        }
+    });
+
+    assert(vol == G->weightedDegree(u, true));
+
+    if (neighborClusterWeights.size() > 0) {
+        // Calculate best cluster
+        index targetCluster = currentCluster;
+        double weightToTargetCluster = weightToCurrent;
+        double bestChange = fitnessChange(u, vol, loop, currentCluster, currentCluster,
+                                          weightToCurrent, weightToCurrent, totalCut);
+        for (index neighborCluster : neighborClusterWeights.insertedIndexes()) {
```

Comment:
I think it would be better to add a method in `SparseVector` that iterates over all the "used" slots (e.g., `forNonZeroElements` or something like that).

Reply:
Do you mean the `forNonZeroElements` function should take a lambda and do the iteration, or should I rather expose a range of iterators?

Reply:
I think that for this case taking a lambda is fine since you don't need to interrupt the loop.

Reply:
Iterators could be added later if needed.